sys/net/gnrc/application_layer/coap/gcoap.c

```diff
@@ -0,0 +1,569 @@
+
+    /* Reserve some space between the header and payload to write options later */
+    pkt->payload      = buf + coap_get_total_hdr_len(pkt) + strlen(path)
+                                                          + GCOAP_REQ_OPTIONS_BUF;
+    /* Payload length really zero at this point, but we set this to the available
+     * length in the buffer. Allows us to reconstruct buffer length later. */
+    pkt->payload_len  = len - (pkt->payload - buf);
+    pkt->content_type = COAP_FORMAT_NONE;
+
+    memcpy(&pkt->url[0], path, strlen(path));
+}
+
+size_t gcoap_finish(coap_pkt_t *pkt, size_t payload_len, unsigned format)
+{
+    /* reconstruct full PDU buffer length */
+    size_t len = pkt->payload_len + (pkt->payload - (uint8_t *)pkt->hdr);
+
+    pkt->content_type = format;
+    pkt->payload_len  = payload_len;
+    return _finish_pdu(pkt, (uint8_t *)pkt->hdr, len);
+}
+
+size_t gcoap_req_send(uint8_t *buf, size_t len, ipv6_addr_t *addr, uint16_t port,
+                                                gcoap_resp_handler_t resp_handler)
+{
+    uint8_t i;
+    gcoap_request_memo_t *memo = NULL;
+
+    /* Find empty slot in list of open requests. */
+    for (i = 0; i < GCOAP_REQ_WAITING_MAX; i++) {
+        if (_coap_state.open_reqs[i].state == GCOAP_MEMO_STATE_UNUSED) {
+            memo = &_coap_state.open_reqs[i];
+            memo->state = GCOAP_MEMO_STATE_WAIT;
+            break;
+        }
+    }
+    if (memo) {
+        memcpy(&memo->hdr_buf[0], buf, GCOAP_HEADER_MAXLEN);
+        memo->resp_handler = resp_handler;
+
+        size_t rv = _send_buf(buf, len, addr, port);
+        if (rv) {
+            /* start response wait timer */
+            memo->timeout_msg.type        = GCOAP_MSG_TYPE_TIMEOUT;
+            memo->timeout_msg.content.ptr = (char *)memo;
+            xtimer_set_msg(&memo->response_timer, GCOAP_RESPONSE_TIMEOUT,
+                                                  &memo->timeout_msg, _pid);
+        } else {
+            memo->state = GCOAP_MEMO_STATE_UNUSED;
+        }
+        return rv;
+    } else {
+        DEBUG("gcoap: dropping request; no space for response tracking\n");
+        return 0;
+    }
+}
+
+void gcoap_resp_init(coap_pkt_t *pkt, uint8_t *buf, size_t len, unsigned code)
+{
+    /* Assume NON type request, so response type is the same. */
+    coap_hdr_set_code(pkt->hdr, code);
+    /* Create message ID since NON? */
+
+    /* Reserve some space between the header and payload to write options later */
+    pkt->payload      = buf + coap_get_total_hdr_len(pkt) + GCOAP_RESP_OPTIONS_BUF;
+    /* Payload length really zero at this point, but we set this to the available
+     * length in the buffer. Allows us to reconstruct buffer length later. */
+    pkt->payload_len  = len - (pkt->payload - buf);
+    pkt->content_type = COAP_FORMAT_NONE;
+}
+
+void gcoap_op_state(uint8_t *open_reqs)
+{
+    uint8_t count = 0;
+    for (uint8_t i = 0; i < GCOAP_REQ_WAITING_MAX; i++) {
```

Comment:
Please use `int` or `unsigned` for iterator variables (cf. http://stackoverflow.com/questions/34891181/for-array-loop-use-int-or-uint8-t).

We should probably add this to the coding conventions.